### ISSUE TITLE
for loops in Julia  more than 6 times slower than in PyPy

### ISSUE THREAD

**Issue by Mirage10** [JuliaLang/julia#13831]
Hello Julia community,

I figured out that in Julia 0.4 a simple 'for loop' takes roughly factor 6 more time than in PyPy 4.0

Julia:
a= 0
for i in 1:1000000000 a+=1 end
=>    13 sec

PyPy:
a=0
for i in range(1000000000): a+=1
=>    < 2 sec

Environment: I am running Julia 0.4 on Linx Mint 17.2 in VMware Virtualization 12.0 + Default option in Julia, nothing special

I am raising this issue, because performance should be first prio in Julia

cheers

**Comment by johnmyleswhite** [comment]
Have you read the perf FAQ? I'm 99% sure this is the global scope issue discussed there.

**Comment by johnmyleswhite** [comment]
See here: http://docs.julialang.org/en/release-0.4/manual/performance-tips/#avoid-global-variables

**Comment by johnmyleswhite** [comment]
For a good example why this is a really dangerous way to try benchmarking Julia:

```jl
julia> function foo()
       a= 0
       for i in 1:1000000000 a+=1 end
       return a
end
foo (generic function with 1 method)

julia> @time foo()
  0.001912 seconds (1.71 k allocations: 90.843 KB)
1000000000

julia> @time foo()
  0.000001 seconds (5 allocations: 176 bytes)
1000000000
```

**Comment by KristofferC** [comment]
To add to that, a gotcha with these type of  "mini" benchmarks in languages that has some sort of compiler is that the  compiler can sometimes optimize away what you are trying to benchmark, ex:

```jl
julia> @code_llvm foo()

define i64 @julia_foo_24661() {
top:
  ret i64 1000000000
}
```

shows that the loop is completely elided because the return value can be inferred.

**Comment by Mirage10** [comment]
@johnmyleswhite : you are absolutely right. I changed my benchmark program a little bit

foo = function()
                 a=0.0
                 for i in 1:1000000000 a+=1.1
              end
              return a
              end

It is a very big difference to initialize with a=0.0 instead of a=0. Also I changed the addition in a+=1.1 otherwise the compiler will do the result directly as @KristofferC  mentioned.

So the last program is 4 time faster in Julia than in PyPy 4.0. Wow. Great achievement!! :-)

So I ran in a typical pitfall at the very first beginning ;-). Maybe this happens to all other newcomer as well. Maybe the compiler could be a little bit more intelligent to detect such situations.
